### PR TITLE
feat: support adding ignored apps for load balancer

### PR DESCRIPTION
**Pull request by Shuo-Jia**
load balancer sometimes expect ignore some app, this pr support the feature

## new remote command

`remote_command -t meta-server meta.lb.ignored_app_list <set|get|clear> [appid1, appid2, ...]
`
## manual test
```
## in bash shell
>>> ./run.sh start_onebox -r 5
>>> ./run.sh stop_onebox_instance -r 1
>>> ./run.sh shell
### in pegasus shell, then create test1 test2
>>> nodes -d
[details]
[details]
address             status    replica_count  primary_count  secondary_count
127.0.0.1:34801  UNALIVE                 0              0                0
127.0.0.1:34802  ALIVE                11              4                7
127.0.0.1:34803  ALIVE                13              4                9
127.0.0.1:34804  ALIVE                12              4                8
127.0.0.1:34805  ALIVE                12              4                8

# here only show some important result
>>> app test1 -d 
[replicas]
pidx  ballot  replica_count  primary             secondaries                              
0     3       3/3            127.0.0.1:34803  [127.0.0.1:34805,127.0.0.1:34804]  
1     3       3/3            127.0.0.1:34805  [127.0.0.1:34802,127.0.0.1:34803]  
2     3       3/3            127.0.0.1:34804  [127.0.0.1:34805,127.0.0.1:34802]  
3     3       3/3            127.0.0.1:34802  [127.0.0.1:34803,127.0.0.1:34805]  
4     3       3/3            127.0.0.1:34803  [127.0.0.1:34804,127.0.0.1:34805]  
5     3       3/3            127.0.0.1:34805  [127.0.0.1:34803,127.0.0.1:34804]  
6     3       3/3            127.0.0.1:34804  [127.0.0.1:34802,127.0.0.1:34803]  
7     3       3/3            127.0.0.1:34802  [127.0.0.1:34804,127.0.0.1:34803]  

[nodes]
node                primary  secondary  total  
127.0.0.1:34802  2        3          5      
127.0.0.1:34803  2        5          7      
127.0.0.1:34804  2        4          6      
127.0.0.1:34805  2        4          6      
                    8        16         24     


>>> app test2 -d 
[replicas]
pidx  ballot  replica_count  primary             secondaries                              
0     3       3/3            127.0.0.1:34802  [127.0.0.1:34805,127.0.0.1:34804]  
1     3       3/3            127.0.0.1:34805  [127.0.0.1:34804,127.0.0.1:34802]  
2     3       3/3            127.0.0.1:34804  [127.0.0.1:34805,127.0.0.1:34803]  
3     3       3/3            127.0.0.1:34803  [127.0.0.1:34802,127.0.0.1:34805]  
4     3       3/3            127.0.0.1:34802  [127.0.0.1:34804,127.0.0.1:34803]  
5     3       3/3            127.0.0.1:34805  [127.0.0.1:34803,127.0.0.1:34804]  
6     3       3/3            127.0.0.1:34804  [127.0.0.1:34803,127.0.0.1:34802]  
7     3       3/3            127.0.0.1:34803  [127.0.0.1:34802,127.0.0.1:34805]  

[nodes]
node                primary  secondary  total  
127.0.0.1:34802  2        4          6      
127.0.0.1:34803  2        4          6      
127.0.0.1:34804  2        4          6      
127.0.0.1:34805  2        4          6      
                    8        16         24     
### exit, then in bash shell
>>> ./run.sh start_onebox_instance -r 1
>>> ./run.sh shell
>>> remote_command -t meta-server meta.lb.ignored_app_list set 1
COMMAND: meta.lb.ignored_app_list set 1

CALL [user-specified] [127.0.0.1:34601] succeed: set ok

>>> set_meta_level lively
### wait a moment and balance finish(`cluster_info` command result the balance count = 0)
>>> nodes -d 
[details]
address             status    replica_count  primary_count  secondary_count
127.0.0.1:34801  ALIVE                 4              1                3
127.0.0.1:34802  ALIVE                10              3                7
127.0.0.1:34803  ALIVE                12              4                8
127.0.0.1:34804  ALIVE                11              4                7
127.0.0.1:34805  ALIVE                11              4                7

>>>app test1 -d # test1 id=1, no balance(no replica move)
[replicas]
pidx  ballot  replica_count  primary             secondaries                              
0     3       3/3            127.0.0.1:34803  [127.0.0.1:34805,127.0.0.1:34804]  
1     3       3/3            127.0.0.1:34805  [127.0.0.1:34802,127.0.0.1:34803]  
2     3       3/3            127.0.0.1:34804  [127.0.0.1:34805,127.0.0.1:34802]  
3     3       3/3            127.0.0.1:34802  [127.0.0.1:34803,127.0.0.1:34805]  
4     3       3/3            127.0.0.1:34803  [127.0.0.1:34804,127.0.0.1:34805]  
5     3       3/3            127.0.0.1:34805  [127.0.0.1:34803,127.0.0.1:34804]  
6     3       3/3            127.0.0.1:34804  [127.0.0.1:34802,127.0.0.1:34803]  
7     3       3/3            127.0.0.1:34802  [127.0.0.1:34804,127.0.0.1:34803]  

[nodes]
node                primary  secondary  total  
127.0.0.1:34802  2        3          5      
127.0.0.1:34803  2        5          7      
127.0.0.1:34804  2        4          6      
127.0.0.1:34805  2        4          6      
                    8        16         24  

>>> app test2 # no ignore, balance(some replica move to 127.0.0.1:34801)
[replicas]
pidx  ballot  replica_count  primary             secondaries                              
0     7       3/3            127.0.0.1:34801  [127.0.0.1:34805,127.0.0.1:34804]  
1     5       3/3            127.0.0.1:34805  [127.0.0.1:34802,127.0.0.1:34801]  
2     5       3/3            127.0.0.1:34804  [127.0.0.1:34803,127.0.0.1:34801]  
3     3       3/3            127.0.0.1:34803  [127.0.0.1:34802,127.0.0.1:34805]  
4     5       3/3            127.0.0.1:34802  [127.0.0.1:34804,127.0.0.1:34801]  
5     3       3/3            127.0.0.1:34805  [127.0.0.1:34803,127.0.0.1:34804]  
6     3       3/3            127.0.0.1:34804  [127.0.0.1:34803,127.0.0.1:34802]  
7     3       3/3            127.0.0.1:34803  [127.0.0.1:34802,127.0.0.1:34805]  

[nodes]
node                primary  secondary  total  
127.0.0.1:34801  1        3          4      
127.0.0.1:34802  1        4          5      
127.0.0.1:34803  2        3          5      
127.0.0.1:34804  2        3          5      
127.0.0.1:34805  2        3          5      
                    8        16         24    

### remote command argument check
>>> remote_command -l 127.0.0.1:34601 meta.lb.ignored_app_list # must contain get/set/clear
COMMAND: meta.lb.ignored_app_list
CALL [user-specified] [127.0.0.1:34601] succeed: invalid arguments

>>> remote_command -l 127.0.0.1:34601 meta.lb.ignored_app_list update # must contain get/set/clear
COMMAND: meta.lb.ignored_app_list update
CALL [user-specified] [127.0.0.1:34601] succeed: invalid arguments

>>> remote_command -l 127.0.0.1:34601 meta.lb.ignored_app_list set # set must have value
COMMAND: meta.lb.ignored_app_list set
CALL [user-specified] [127.0.0.1:34601] succeed: invalid arguments

>>> remote_command -l 127.0.0.1:34601 meta.lb.ignored_app_list set 1,2
COMMAND: meta.lb.ignored_app_list set 1,2
CALL [user-specified] [127.0.0.1:34601] succeed: set ok

>>> remote_command -l 127.0.0.1:34601 meta.lb.ignored_app_list get
COMMAND: meta.lb.ignored_app_list get
CALL [user-specified] [127.0.0.1:34601] succeed: ignored_app_id_list: 1,2

>>> >>> remote_command -l 127.0.0.1:34601 meta.lb.ignored_app_list clear
COMMAND: meta.lb.ignored_app_list clear
CALL [user-specified] [127.0.0.1:34601] succeed: clear ok

>>> remote_command -l 127.0.0.1:34601 meta.lb.ignored_app_list get
COMMAND: meta.lb.ignored_app_list get
CALL [user-specified] [127.0.0.1:34601] succeed: no ignored apps
```

### CODE REVIEW

File: src/dist/replication/meta_server/greedy_load_balancer.cpp [hunk 1]
```diff
@@ -945,5 +964,56 @@ void greedy_load_balancer::report(const dsn::replication::migration_list &list,
         _recent_balance_copy_secondary_count->add(counters[COPY_SEC_COUNT]);
     }
 }
+
+std::string
+greedy_load_balancer::set_balancer_ignored_app_id_list(const std::vector<std::string> &args)
+{
+    std::string invalid_arguments("invalid arguments");
```
Comment (neverchanje) [inline src/dist/replication/meta_server/greedy_load_balancer.cpp:971]:
```suggestion
    static const std::string invalid_arguments("invalid arguments");
```

File: src/dist/replication/meta_server/greedy_load_balancer.cpp [hunk 2]
```diff
@@ -945,5 +964,56 @@ void greedy_load_balancer::report(const dsn::replication::migration_list &list,
...
+    std::stringstream oss;
+    if (args.empty()) {
+        dsn::zauto_read_lock l(_balancer_ignored_apps_lock);
+        oss << "ignored_app_id_list: ";
+        std::copy(_balancer_ignored_apps.begin(),
+                  _balancer_ignored_apps.end(),
+                  std::ostream_iterator<dsn::replication::app_id>(oss, ","));
```
Comment (neverchanje) [inline src/dist/replication/meta_server/greedy_load_balancer.cpp:978]:
```suggestion
                  std::ostream_iterator<app_id>(oss, ","));
```

File: src/dist/replication/meta_server/greedy_load_balancer.cpp [hunk 3]
```diff
@@ -945,5 +964,56 @@ void greedy_load_balancer::report(const dsn::replication::migration_list &list,
...
+        std::string app_ids = oss.str();
+        app_ids.erase(app_ids.size() - 1);
```
Comment (neverchanje) [inline src/dist/replication/meta_server/greedy_load_balancer.cpp:980]: erase may be heavy, because internally it may be implemented by memove. 
```suggestion
        app_ids[app_ids.size() - 1] = '\0';
```

File: src/dist/replication/meta_server/greedy_load_balancer.cpp [hunk 4]
```diff
@@ -945,5 +964,56 @@ void greedy_load_balancer::report(const dsn::replication::migration_list &list,
...
+    }
+
+    std::vector<std::string> app_ids;
+    dsn::utils::split_args(args[0].c_str(), app_ids, ',');
+    if (args.size() == 0) {
+        return invalid_arguments;
+    }
+
+    std::set<dsn::replication::app_id> app_list;
+    for (const std::string &app_id : app_ids) {
```
Comment (neverchanje) [inline src/dist/replication/meta_server/greedy_load_balancer.cpp:1001]:
```suggestion
    for (const std::string &app_id_str : app_ids) {
```

File: src/dist/replication/meta_server/greedy_load_balancer.cpp [hunk 5]
```diff
@@ -945,5 +964,56 @@ void greedy_load_balancer::report(const dsn::replication::migration_list &list,
...
+        app_list.insert(app);
+    }
+    _balancer_ignored_apps = std::move(app_list);
+    return "set ok";
+}
+
+bool greedy_load_balancer::in_ignored_apps(dsn::replication::app_id app_id)
+{
+    dsn::zauto_read_lock l(_balancer_ignored_apps_lock);
+    return _balancer_ignored_apps.count(app_id) != 0;
```
Comment (neverchanje) [inline src/dist/replication/meta_server/greedy_load_balancer.cpp:1015]:
```suggestion
    return _balancer_ignored_apps.find(app_id) != _balancer_ignored_apps.end();
```

File: src/dist/replication/meta_server/greedy_load_balancer.cpp [hunk 6]
```diff
@@ -819,6 +829,10 @@ void greedy_load_balancer::greedy_balancer(const bool balance_checker)
 
     for (const auto &kv : apps) {
         const std::shared_ptr<app_state> &app = kv.second;
+        if (in_ignored_apps(kv.first)) {
+            ddebug_f("stop to do primary balance for the ignored app[{}]", app->get_logname());
```
Comment (acelyc111) [inline src/dist/replication/meta_server/greedy_load_balancer.cpp:833]:
```suggestion
            ddebug_f("skip to do primary balance for the ignored app[{}]", app->get_logname());
```

File: src/dist/replication/meta_server/greedy_load_balancer.cpp [hunk 7]
```diff
@@ -861,6 +875,11 @@ void greedy_load_balancer::greedy_balancer(const bool balance_checker)
     // another app's primary balancer if not seperated.
     for (const auto &kv : apps) {
         const std::shared_ptr<app_state> &app = kv.second;
+        if (in_ignored_apps(kv.first)) {
+            ddebug_f("stop to do secondary balance for the ignored app[{}]", app->get_logname());
```
Comment (acelyc111) [inline src/dist/replication/meta_server/greedy_load_balancer.cpp:879]:
```suggestion
            ddebug_f("skip to do secondary balance for the ignored app[{}]", app->get_logname());
```

File: src/dist/replication/meta_server/greedy_load_balancer.cpp [hunk 8]
```diff
@@ -945,5 +964,56 @@ void greedy_load_balancer::report(const dsn::replication::migration_list &list,
...
+    if (args.empty()) {
+        dsn::zauto_read_lock l(_balancer_ignored_apps_lock);
+        oss << "ignored_app_id_list: ";
+        std::copy(_balancer_ignored_apps.begin(),
+                  _balancer_ignored_apps.end(),
+                  std::ostream_iterator<app_id>(oss, ","));
+        std::string app_ids = oss.str();
+        app_ids[app_ids.size() - 1] = '\0';
+        return app_ids;
+    }
```
Comment (acelyc111) [inline src/dist/replication/meta_server/greedy_load_balancer.cpp:982]: Empty `args ` means get _balancer_ignored_apps?

File: src/dist/replication/meta_server/greedy_load_balancer.cpp [hunk 9]
```diff
@@ -945,5 +964,56 @@ void greedy_load_balancer::report(const dsn::replication::migration_list &list,
...
+    std::set<app_id> app_list;
+    for (const std::string &app_id_str : app_ids) {
+        app_id app;
+        if (!dsn::buf2int32(app_id_str, app)) {
+            return invalid_arguments;
+        }
+        app_list.insert(app);
+    }
+    _balancer_ignored_apps = std::move(app_list);
+    return "set ok";
```
Comment (hycdong) [inline src/dist/replication/meta_server/greedy_load_balancer.cpp:1009]: - if `args` is empty, get all ignored app id, otherwise, `args` length will always be equal to 1
- if `args` = clear, clear all ignored app id, otherwise, it should be like `app_id1, app_id2 ...`

I think your design is just like other remote command, like `replica.mem-release-max-reserved-percentage`, no args means get the value, has args means set the value. However, it will be strange that you can get the value from a setter function. Besides, you combine set and clear in this function, but your remote command explanation doesn't show this usage `lb.ignored_app_list [app_id1,app_id2..]`, I don't you how to clear ignored list unless reading source code.

In conclusion, I suggest you separate getter, setter into two function, add a new remote command called `lb.query_ignored_app`, you can return app_id and app_name to make it more clearer. For original setter function, no args mean clear ignore app id,  has args means set ignore app id. If you insist on combining three functions into one function, rename the function, add more comments and fulfill remote command usage explanation.

**Comment by acelyc111** [comment]
>
 >>>app test1 -d # test1 id=6, no balance(no replica move to 127.0.0.1:34801)
[nodes]
node                primary  secondary  total  
127.0.0.1:34802  2        4          6      
127.0.0.1:34803  2        4          6      
127.0.0.1:34804  2        4          6      
127.0.0.1:34805  2        4          6      
>>> app test2 -d # test2 id=7, no balance(no replica move to 127.0.0.1:34801)
[nodes]
node                primary  secondary  total  
127.0.0.1:34802  4        8          12     
127.0.0.1:34803  4        8          12     
127.0.0.1:34804  4        9          13     
127.0.0.1:34805  4        7          11     

The output is not command of `app test1 -d`, and you should prove that replicas in nodes 02~05 are not moved internally though replica count are not changed.
Besides, give 1 example is enough.

File: src/dist/replication/meta_server/greedy_load_balancer.cpp [hunk 10]
```diff
@@ -119,6 +120,18 @@ void greedy_load_balancer::register_ctrl_commands()
...
+        {"lb.ignored_app_list"},
+        "lb.ignored_app_list [app_id1,app_id2.. | CLEAR](if no argument means get current ignored "
+        "list)",
+        "get or update balance ignored_app_list",
+        [this](const std::vector<std::string> &args) {
+            if (args.empty()) {
+                return get_balancer_ignored_app_ids();
+            }
+            return update_balancer_ignored_app_ids(args);
+        });
```
Comment (acelyc111) [inline src/dist/replication/meta_server/greedy_load_balancer.cpp:134]: The command semantics seem a bit of strange, how about:
lb.ignored_app_list get  # to get ignored app ids
lb.ignored_app_list set (x,x,x) # to set ignored app ids
lb.ignored_app_list clear  # to clear ignored app ids

File: src/dist/replication/meta_server/greedy_load_balancer.cpp [hunk 11]
```diff
@@ -119,6 +120,18 @@ void greedy_load_balancer::register_ctrl_commands()
         "lb.get_balance_operation_count [total | move_pri | copy_pri | copy_sec | detail]",
         "get balance operation count",
         [this](const std::vector<std::string> &args) { return get_balance_operation_count(args); });
+
+    _ctrl_balancer_ignored_apps = dsn::command_manager::instance().register_app_command(
+        {"lb.ignored_app_list"},
+        "lb.ignored_app_list [app_id1,app_id2.. | CLEAR](if no argument means get current ignored "
+        "list)",
+        "get or update balance ignored_app_list",
```
Comment (levy5307) [inline src/dist/replication/meta_server/greedy_load_balancer.cpp:128]: get/update/clear

File: src/dist/replication/meta_server/greedy_load_balancer.cpp [hunk 12]
```diff
@@ -819,6 +833,10 @@ void greedy_load_balancer::greedy_balancer(const bool balance_checker)
 
     for (const auto &kv : apps) {
         const std::shared_ptr<app_state> &app = kv.second;
+        if (in_ignored_apps(kv.first)) {
```
Comment (levy5307) [inline src/dist/replication/meta_server/greedy_load_balancer.cpp:836]: I think you can move this code behind of `if (app->status != app_status::AS_AVAILABLE)`

File: src/dist/replication/meta_server/greedy_load_balancer.cpp (hunk 10 above)
Comment (Shuo-Jia) [reply src/dist/replication/meta_server/greedy_load_balancer.cpp:134]: This is  coincident with other `remote command` : `empty` means `get current value`, `valid argument` means `set currnet value`, `clear or default`  means recovery, I think this rule may no need change.

Comment (acelyc111) [reply src/dist/replication/meta_server/greedy_load_balancer.cpp:134]: > lb.ignored_app_list get # to get ignored app ids
lb.ignored_app_list set x,x,x # to set ignored app ids
lb.ignored_app_list clear # to clear ignored app ids

It's a simple semantics rule to obey, `<topic> <action> [parameters]`.
I think it's not needed to keep with other commands, and we could update old commands if we found a better solution.

**Comment by Shuo-Jia** [comment]
> > > > app test1 -d # test1 id=6, no balance(no replica move to 127.0.0.1:34801)
> > > > [nodes]
> > > > node                primary  secondary  total
> > > > 127.0.0.1:34802  2        4          6
> > > > 127.0.0.1:34803  2        4          6
> > > > 127.0.0.1:34804  2        4          6
> > > > 127.0.0.1:34805  2        4          6
> > > > app test2 -d # test2 id=7, no balance(no replica move to 127.0.0.1:34801)
> > > > [nodes]
> > > > node                primary  secondary  total
> > > > 127.0.0.1:34802  4        8          12
> > > > 127.0.0.1:34803  4        8          12
> > > > 127.0.0.1:34804  4        9          13
> > > > 127.0.0.1:34805  4        7          11
> 
> The output is not command of `app test1 -d`, and you should prove that replicas on nodes 02~05 are not moved internally though total replica count are not changed.
> Besides, giving one example is enough.

have update `manul test` description

File: src/dist/replication/meta_server/greedy_load_balancer.cpp [hunk 13]
```diff
@@ -945,5 +964,77 @@ void greedy_load_balancer::report(const dsn::replication::migration_list &list,
...
+    if (args[0] == "clear") {
+        return clear_balancer_ignored_app_ids();
+    }
+    return invalid_arguments;
+}
+
+std::string greedy_load_balancer::set_balancer_ignored_app_ids(const std::vector<std::string> &args)
+{
+    static const std::string invalid_arguments("invalid arguments");
+    dsn::zauto_write_lock l(_balancer_ignored_apps_lock);
```
Comment (acelyc111) [inline src/dist/replication/meta_server/greedy_load_balancer.cpp:990]: Move this line before `_balancer_ignored_apps = std::move(app_list);` to reduce lock scope.

File: src/dist/replication/meta_server/greedy_load_balancer.cpp [hunk 14]
```diff
@@ -945,5 +964,78 @@ void greedy_load_balancer::report(const dsn::replication::migration_list &list,
...
+}
+
+std::string greedy_load_balancer::clear_balancer_ignored_app_ids()
+{
+    dsn::zauto_write_lock l(_balancer_ignored_apps_lock);
+    _balancer_ignored_apps.clear();
+    return "clear ok";
+}
+
+bool greedy_load_balancer::in_ignored_apps(app_id app_id)
```
Comment (levy5307) [inline src/dist/replication/meta_server/greedy_load_balancer.cpp:1034]: I think `is_ignored_app` is a little better.
